Document how Positron discovers available R installations (#21)
posit-dev/positron-website: r-installations.qmd

---

> title: "Discovering R installations"

Reviewer: ```suggestion
title: "Discovering R Installations"
```

---

:::{.callout-note}
This information applies to the **Public Beta** release of Positron.
The rules governing interpreter discovery are under active development.
This guide is intended to help you understand the current behavior.
:::

This guide pertains to the R versions that appear in the top-right interpreter chooser (also known as the _Top Bar Interpreter Dropdown_ internally).

## R installation discovery

Positron consults various sources to build the list of available R interpreters:

* Well-known R root folders for the specific operating system:
  - macOS: `/Library/Frameworks/R.framework/Versions`
  - Linux: `/opt/R`

>   - Windows: `%ProgramW6432%/R` (typically resolves to `C:/Program Files/R`) and
>     `%LOCALAPPDATA%/Programs/R` (typically resolves to
>     `C:/Users/username/AppData/Local/Programs/R`)
>   These locations are treated as so-called root folders, meaning we expect them
>   to _contain_ 1 or more versions of R, not to actually _be_ an R installation.

Reviewer: I'm finding how this line is formatted confusing. Does it really belong just to the Windows list item? If so, maybe a semicolon?

```suggestion
  - Windows: `%ProgramW6432%/R` (typically resolves to `C:/Program Files/R`) and
    `%LOCALAPPDATA%/Programs/R` (typically resolves to
    `C:/Users/username/AppData/Local/Programs/R`); these locations are treated as so-called root folders, meaning we expect them
  to _contain_ 1 or more versions of R, not to actually _be_ an R installation.
```

Reviewer: Or does it belong to all three? I think we need some formatting adjustment:

![Screenshot 2024-11-20 at 1 52 04 PM](https://github.com/user-attachments/assets/f3667039-9582-46ba-a92e-5b5daabe16d5)

Author: This is sorted out.

* The `PATH`, in the sense of the R executable returned by `which R`
* Well-known paths to an R executable or to a symlink that resolves to an R executable, e.g.:
  - `/usr/bin/R`
  - `/usr/local/bin/R`
  - `/opt/local/bin/R`
  - `/opt/homebrew/bin/R`
* R installations recorded in the Windows registry, in keys such as:
  - `HKEY_CURRENT_USER\SOFTWARE\R-core\R64\X.Y.Z`
  - `HKEY_LOCAL_MACHINE\SOFTWARE\R-core\R64\X.Y.Z`
  where `X.Y.Z` is a concrete R version, such as 4.4.2.
  - It's often possible to discover which of those R versions is considered
    the current version of R from the registry keys
    `HKEY_CURRENT_USER\SOFTWARE\R-core\R64\InstallPath` or
    `HKEY_LOCAL_MACHINE\SOFTWARE\R-core\R64\InstallPath`.

## Customizing R discovery

If you have an R installation that won't be discovered by the search strategy described above, two settings are available to instruct Positron to look in additional locations:

* [`positron.r.customRootFolders`](positron://settings/positron.r.customRootFolders): Suitable for R root folders that contain 1 or more R installations.
  - Hypothetical example: `C:/nonstandardRLocation`

> * [`positron.r.customBinaries`](positron://settings/positron.r.customBinaries): Suitable for specific R executables.

Reviewer: This is a fairly surprising way to use a link on a website. @AshleyHenry15 do you mind giving some input on this kind of choice in documentation?

If someone has Positron installed, it opens directly to the setting -- so cool! If someone does not have Positron installed or is on mobile, it errors, which may be quite surprising given that it looks like a link to click on:

![Discovering R installations – Positron](https://github.com/user-attachments/assets/0c4b90a8-369f-42ac-a48f-f1035214c1fb)

Reviewer 2: Yes - I agree; this isn't the normal behavior for our docs, so we shouldn't add these kinks as such. People would likely click on them expecting it to redirect them to another documentation page. I also don't know how I feel about clicking something that opens software without someone understanding what is going to happen before doing it.

---
title: "Discovering R Installations"
---

:::{.callout-note}
This information applies to the **Public Beta** release of Positron.
The rules governing interpreter discovery are under active development.
This guide is intended to help you understand the current behavior.

> :::

Reviewer 2: Since we use this in two different files - this should be added as an include in the `_snippets` directory.

Author: The text is actually different in these 2 callouts. So to do this we'd also have to use the exact same text and make it generic. Is that what we want?

Reviewer: No, let's use the different text in these two places and not use a snippet.

> This guide pertains to the R versions that appear in the top-right interpreter chooser (also known as the _Top Bar Interpreter Dropdown_ internally).

Reviewer 2: ```suggestion
This guide pertains to the R versions in the top-right interpreter chooser (internally known as the Top Bar Interpreter dropdown.
```

## R installation discovery

Positron consults various sources to build the list of available R interpreters:

> * Well-known R root folders for the specific operating system:

Reviewer 2: ```suggestion
- Well-known R root folders for the specific operating system:
```

Author: The top-level bullets are done with `*` and the second level with `-`, so this change alone doesn't make sense.

Reviewer: Yes, I agree.

Author: OK I'm going to resolve the conversations about changing `*` to `-` (without changing them).

  - macOS: `/Library/Frameworks/R.framework/Versions`
  - Linux: `/opt/R`
  - Windows: `%ProgramW6432%/R` (typically resolves to `C:/Program Files/R`) and
    `%LOCALAPPDATA%/Programs/R` (typically resolves to
    `C:/Users/username/AppData/Local/Programs/R`)

  These locations are treated as so-called root folders, meaning we expect them

>   to _contain_ 1 or more versions of R, not to actually _be_ an R installation.

Reviewer 2: There isn't a line blank line between the list item and this statement in the rendered docs.

Author: The blank line is necessary to make this not become part of the second subbullet.

>   These locations are treated as so-called root folders, meaning we expect them
>   to _contain_ 1 or more versions of R, not to actually _be_ an R installation.

Reviewer 2: ```suggestion
  These locations are treated as so-called root folders, meaning we expect them
  to contain one or more versions of R, not to actually be an R installation.
```

> * The `PATH`, in the sense of the R executable returned by `which R`
> * Well-known paths to an R executable or to a symlink that resolves to an R executable, e.g.:

Reviewer 2: ```suggestion
- The `PATH`, in the sense of the R executable returned by `which R`
- Well-known paths to an R executable or to a symlink that resolves to an R executable, e.g.:
```

  - `/usr/bin/R`
  - `/usr/local/bin/R`
  - `/opt/local/bin/R`
  - `/opt/homebrew/bin/R`

> * R installations recorded in the Windows registry, in keys such as:

Reviewer 2: ```suggestion
- R installations recorded in the Windows registry, in keys such as:
```

  - `HKEY_CURRENT_USER\SOFTWARE\R-core\R64\X.Y.Z`
  - `HKEY_LOCAL_MACHINE\SOFTWARE\R-core\R64\X.Y.Z`
  where `X.Y.Z` is a concrete R version, such as 4.4.2.
  - It's often possible to discover which of those R versions is considered
    the current version of R from the registry keys
    `HKEY_CURRENT_USER\SOFTWARE\R-core\R64\InstallPath` or
    `HKEY_LOCAL_MACHINE\SOFTWARE\R-core\R64\InstallPath`.

## Customizing R discovery

If you have an R installation that won't be discovered by the search strategy described above, two settings are available to instruct Positron to look in additional locations:

> * [`positron.r.customRootFolders`](positron://settings/positron.r.customRootFolders): Suitable for R root folders that contain 1 or more R installations.

Reviewer 2: ```suggestion
- `positron.r.customRootFolders`: Suitable for R root folders that contain 1 or more R installations.
```

> * [`positron.r.customRootFolders`](positron://settings/positron.r.customRootFolders): Suitable for R root folders that contain 1 or more R installations.
>   - Hypothetical example: `C:/nonstandardRLocation`
> * [`positron.r.customBinaries`](positron://settings/positron.r.customBinaries): Suitable for specific R executables.

Reviewer 2: ```suggestion
- `positron.r.customRootFolders`: Suitable for R root folders that contain one or more R installations.
  - Hypothetical example: `C:/nonstandardRLocation`
- `positron.r.customBinaries`: Suitable for specific R executables.
```

  - Hypothetical example: `C:/nonstandardRLocation/R-4.4.1/bin/x64/R.exe`

> One way to navigate to these settings is to use the command palette to do *Preferences: Open Settings*.

Reviewer 2: ```suggestion
One way to navigate to these settings is to use the command palette to open Preferences: Open Settings.
```

Reviewer 2: If you want to instruct the user to type that exactly within the command pallet, then you can use type "Preferences: Open Settings"

> Then go to *Extensions > R > Positron R Language Pack (advanced)*.

Reviewer 2: ```suggestion
Then navigate to **Extensions** > **R** > **Positron R Language Pack (advanced)**.
```

> It's important to use concrete paths in these settings and not rely on shell features, such as tilde (`~/`) or parameter (`${HOME}`) expansion.

Reviewer 2: ```suggestion
It's important to use absolute paths in these settings and not rely on shell features, such as tilde (`~/`) or parameter (`${HOME}`) expansion.
```

## Unsupported R installations

Positron only supports R 4.2 and higher.
Older R versions might be discovered in the search described above, but they will be discarded.

> On macOS, Positron only supports the system's "current" R version and other R installations that are **orthogonal**.

Reviewer 2: ```suggestion
On macOS, Positron only supports the system's current R version and other R installations that are orthogonal.
```

If "current' isn't completely accurate, you could add a footnote here to add additional information.

Author: It's more that "current" can mean different things to different people, so this is really "current for some definition of current". And I'm not going to belabor it here. But I will remove the quotes in the name of moving things along.

> What does orthogonal mean here?

Reviewer 2: I would add a line break or something here - this feels a bit awkward nested within the paragraph.

The `R` executable on macOS (and Linux) is actually a shell script that's responsible for launching R.
In an orthogonal R installation, this script targets the home of a specific, concrete version of R, such as `/Library/Frameworks/R.framework/Versions/4.4-arm64`.

> In a non-orthogonal R installation, this script targets the home of "the current R version" (literally, `/Library/Frameworks/R.framework/Resources`, which is a symlink).

Reviewer 2: ```suggestion
In a non-orthogonal R installation, this script targets the home of the current R version (literally, `/Library/Frameworks/R.framework/Resources`, which is a symlink).
```

The CRAN installer leaves a non-orthogonal R installation behind and, by default, removes the previous version of R.

The notion of orthogonality comes from a tool called [rig: The R Installation Manager](https://github.com/r-lib/rig).

> We highly recommend rig for managing R installation, especially for macOS users who would like to have more than 1 functioning R installation.

Reviewer 2: ```suggestion
We highly recommend rig for managing R installation, especially for macOS users who want to have more than one functioning R installation.
```

## Troubleshooting

> The discovery process is logged in the "Positron R Extension" channel.

Reviewer 2: ```suggestion
The discovery process is logged in the Positron R Extension channel.
```
